nixos/modules/services/misc/ollama.nix

```diff
@@ -255,14 +260,44 @@ in
           ];
           UMask = "0077";
         };
-      postStart = mkBefore ''
-        set -x
-        export OLLAMA_HOST=${lib.escapeShellArg cfg.host}:${builtins.toString cfg.port}
-        for model in ${lib.escapeShellArgs cfg.loadModels}
-        do
-          ${lib.escapeShellArg (lib.getExe ollamaPackage)} pull "$model"
-        done
-      '';
+    };
+
+    systemd.services.ollama-model-loader = lib.mkIf ((builtins.length cfg.loadModels) > 0) {
+      description = "Download ollama models in the background";
+      wantedBy = [
+        "multi-user.target"
+        "ollama.service"
+      ];
+      after = [ "ollama.service" ];
+      bindsTo = [ "ollama.service" ];
+      environment = config.systemd.services.ollama.environment;
+      serviceConfig = {
+        Type = "exec";
+        DynamicUser = true;
+        Restart = "on-failure";
+        # bounded exponential backoff
+        RestartSec = "1s";
+        RestartMaxDelaySec = "2h";
+        RestartSteps = "10";
+      };
+
+      script =
+        let
+          loadModels = builtins.map (
+            model: "'${lib.getExe ollamaPackage}' pull ${lib.escapeShellArg model} &"
+          ) cfg.loadModels;
+        in
+        ''
+          ${lib.concatStringsSep "\n" loadModels}
+          for job in $(jobs -p); do
+            wait $job
+            exit_code="$?"
+
+            if [ "$exit_code" != 0 ]; then
+              exit "$exit_code"
+            fi
+          done
```

Comment:
I don't really understand this part. First time I see the `jobs` command.

Also, I'm wondering if there would be an easier way to do that. I have the feeling that this is overly complex... or maybe it's just because I'm surprised to see this implementation.

Reply:
> I don't really understand this part. First time I see the jobs command.

A bash-ism for when use spawn subprocesses using `cmd &`.
This is nice compared to the previous state, and I wouldn't mind merging this, but if we're discussing this bit, I'd say:

- This could get the user rate-limited by upstream, since all of these `pull`s start simultaneously without any batching
- If the first job fails, all the others are cancelled immediately? Or are they? Depends on how systemd does the clean-up, since there's no `trap`


If we're looking for alternatives, one would be to make this a template job, and make ollama depend (`wants`) on a list of these template services. If we had to escape the names, we could e.g. generate a json mapping the escaped names into full URIs (or whatever ollama uses) and have the template unit convert the escaped name using json and pull that single model. This way they're cancelled independently, the progress is preserved, and we're also avoiding the bash magic in favour of systemd magic.

Personally, I'd be much more interested in using static models as FODs, but I'm not currently using ollama

Reply:
Thank you, very interesting.

In that case, as an alternative, wouldn't it be cleaner to create one Systemd service per download?

Reply:
`jobs -p` returns the process IDs of all currently running background jobs. The whole loop is specifically there to propagate any failures, since just `wait` will return success even if a child process fails. When the service fails (perhaps due to no internet connection), it will be periodically restarted until it succeeds. Without this loop (using just `wait`), the service always succeeds, which prevents this automatic restarting from working.

Reply:
> This could get the user rate-limited by upstream, since all of these `pull`s start simultaneously without any batching

I would think that the ollama service itself would manage these connections in a way that's compatible with the model servers, since both are controlled by the same developers? This probably needs investigation, though.

> If we're looking for alternatives, one would be to make this a template job

An interesting idea, I'll look into it. I'm not familiar with template units, so it may take a while.

> static models as FODs

I don't know what `static model` or `FOD` refers to; could you explain?

Reply:
> ollama service itself would manage these connections in a way that's compatible with the model servers, since both are controlled by the same developers?

Oh, are you saying that is handled by a daemon, rather than by the processes spawned by `ollama pull`?

> I don't know what static model or FOD refers to; could you explain?

I meant weights downloaded into the store using a fixed-ouput derivation (fod). I think we brought this up before maybe at https://gist.github.com/309d4e26fdf0af328968cfcfac33347e#file-llama-2-7b-nix-L43-L53

Reply:
> Oh, are you saying that is handled by a daemon, rather than by the processes spawned by `ollama pull`?

Yes, the models are downloaded by the service (`ollama serve`); `ollama pull` just tells the service to download a model via http request. However, I don't necessarily know if the ollama server implements any safety checks to prevent rate-limiting.

> I meant weights downloaded into the store using a fixed-ouput derivation (fod).

Ah, yes; I do think that would be nicer, but it seems quite difficult given that the model can potentially change at any time. I'm not sure if there's any practical way to make it work, unless I'm misunderstanding how or where the models are being hosted.

As far as I understand, the model files are hosted on ollama's web servers, and only the most recent version is kept. That doesn't play well with fixed-output derivations. And it's probably not worth re-hosting all of the models just to be able to keep old versions around, given how large they are.

Reply:
Well at least the huggingface stuff is git and you just pin a revision. But this is fine, people who want the static set-up and people who want upstream ollama's dynamic workflow are different audiences, and this `loadModels` option is there to accommodate the latter I guess

Reply:
> In that case, as an alternative, wouldn't it be cleaner to create one Systemd service per download?

Would this be a bad idea?

Reply:
> one Systemd service per download

I don't think either option is strictly better or worse; there are multiple tradeoffs to switching to multiple units or template units.

Pros:
- individual model downloads can fail and restart individually, potentially for separate reasons (eg network failure and invalid model name)
- logs would be isolated between a particular model and the unit that downloaded it
- no bash script that could be hiding obscure bugs (though bash isn't technically the only option)

Cons:
- it's difficult to restart all model downloads at once, since they are split among multiple different units
  - may not be a problem if support is added for timers to automatically update models periodically
- what happens if the user passes a string containing characters that are illegal in filenames?
  - should a check be added to verify that only legal characters are used?
  - should any characters be substituted in the unit name (eg replace `:` with `-`)?
  - is there some practical way to check the requested model against all possible models?

Overall, I think multiple units is slightly better, but has relatively minor additional benefits for a disproportionate amount of time and effort. It seems to me that it may be better to merge now, immediately improving user experience, and leave this change for another pull request.

Reply:
Fair enough, you have might point of view.

If you think this is the best option, then fine for me.